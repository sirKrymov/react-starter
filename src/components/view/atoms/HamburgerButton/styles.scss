@import '../../../../styles/configs';

.hamburger-btn {
  background-color: transparent;
  position: relative;
  display: block;
  outline: none;
  padding: 0;
  cursor: pointer;
  border: 0;
  height: 24px;
  width: 24px;

  p {
    background-color: var(--secondary);
    transition: $anim-time ease;
    transform: rotate(0deg);
    position: absolute;
    display: block;
    opacity: 1;
    height: 3px;
    width: 100%;

    &:nth-of-type(1) {
      left: 0;
      top: 0;
    }

    &:nth-of-type(2) {
      right: 0;
      top: 10px;
    }

    &:nth-of-type(3) {
      left: 0;
      top: 20px;
    }
  }

  &--open {
    p {
      background-color: var(--secondaryDark);

      &:nth-of-type(1) {
        transform: rotate(135deg);
        top: 10px;
      }

      &:nth-of-type(2) {
        opacity: 0;
        right: 26px;
      }

      &:nth-of-type(3) {
        transform: rotate(-135deg);
        top: 10px;
      }
    }
  }

  &--mirror {
    p {
      &:nth-of-type(1) {
        right: 0;
        left: initial;
        top: 0;
      }

      &:nth-of-type(2) {
        right: initial;
        left: 0;
        top: 10px;
      }

      &:nth-of-type(3) {
        right: 0;
        left: initial;
        top: 20px;
      }
    }
  }

  &--mirror-open {
    p {
      background-color: var(--secondaryDark);

      &:nth-of-type(1) {
        transform: rotate(-135deg);
        top: 10px;
      }

      &:nth-of-type(2) {
        opacity: 0;
        right: 0;
        left: 26px;
      }

      &:nth-of-type(3) {
        transform: rotate(135deg);
        top: 10px;
      }
    }
  }
}
