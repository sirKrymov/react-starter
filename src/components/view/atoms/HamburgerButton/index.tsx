// @SETUP: Please don't forget to remove this component or current comment if component is used in project
import React, { ReactElement } from 'react';

import cn from 'classnames';

import './styles.scss';

interface IProps {
  onClick(): void;
  mirror?: boolean;
  open?: boolean;
  id: string;
}

export function HamburgerButton({
  onClick,
  mirror,
  open = false,
  id
}: IProps): ReactElement {
  const hamburgerButtonClass = cn({
    'hamburger-btn': true,
    'hamburger-btn--mirror-open': mirror && open,
    'hamburger-btn--mirror': mirror,
    'hamburger-btn--open': open
  });

  return (
    <button className={hamburgerButtonClass} onClick={onClick} id={id}>
      <p />
      <p />
      <p />
    </button>
  );
}
