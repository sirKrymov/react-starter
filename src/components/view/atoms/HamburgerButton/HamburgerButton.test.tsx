import React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';

import { HamburgerButton } from '../HamburgerButton';

describe('<HamburgerButton> atom component', () => {
  const handleClick = sinon.spy();
  let shallowComponent: ShallowWrapper;

  const createComponent = (props?: Record<string, any>) => {
    shallowComponent = shallow(
      <HamburgerButton onClick={handleClick} id="hamburger" {...props} />
    );
  };

  describe('WHEN the component loads', () => {
    beforeEach(() => {
      createComponent();
    });

    it('THEN it should render without crashing', () => {
      expect(shallowComponent.html()).not.toHaveLength(0);
    });

    it('THEN it should match snapshot', () => {
      expect(shallowToJson(shallowComponent)).toMatchSnapshot();
    });

    it('AND it should handle initial button logic', () => {
      shallowComponent.find('button').simulate('click');
      expect(handleClick).toHaveProperty('callCount', 1);
    });
  });

  describe('WHEN the component have additional properties', () => {
    it('AND it should handle initial button logic', () => {
      createComponent({ open: true });

      expect(shallowComponent.find('button.hamburger-btn--open')).toHaveLength(
        1
      );
    });

    it('AND it should handle initial button logic', () => {
      createComponent({ open: false });

      expect(shallowComponent.find('button.hamburger-btn--open')).toHaveLength(
        0
      );
    });

    it('AND it should handle click logic', () => {
      createComponent({ mirror: true, open: true });

      expect(
        shallowComponent.find('button.hamburger-btn--mirror-open')
      ).toHaveLength(1);
    });
  });
});
