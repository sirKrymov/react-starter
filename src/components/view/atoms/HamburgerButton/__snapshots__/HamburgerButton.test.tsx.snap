// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HamburgerButton> atom component WHEN the component loads THEN it should match snapshot 1`] = `
<button
  className="hamburger-btn"
  id="hamburger"
  onClick={[Function]}
>
  <p />
  <p />
  <p />
</button>
`;
